File: gke_customization/gke_catalog/report/grade_&_designation_wise_employee_count/grade_&_designation_wise_employee_count.py
# Copyright (c) 2025, Gurukrupa Export
# For license information, please see license.txt

import frappe
from frappe import _


# -------------------------------------------------------
# Execute
# -------------------------------------------------------

def execute(filters=None):
    filters = filters or {}

    columns = get_columns()
    data = get_data()
    message = get_message()

    return columns, data, message


# -------------------------------------------------------
# Columns
# -------------------------------------------------------

def get_columns():
    return [
        {
            "label": _("Employee Grade"),
            "fieldname": "employee_grade",
            "fieldtype": "Data",
            "width": 100,
        },
        {
            "label": _("Number of Employees"),
            "fieldname": "number_of_employees",
            "fieldtype": "Int",
            "width": 100,
        },
        {
            "label": _("Designations"),
            "fieldname": "designations",
            "fieldtype": "Data",
            "width": 1080,
        },
        
    ]


# -------------------------------------------------------
# Data
# -------------------------------------------------------

def get_data():
    query = """
        SELECT
            eg.name AS employee_grade,
            GROUP_CONCAT(
                DISTINCT d.name
                ORDER BY d.name
                SEPARATOR ', '
            ) AS designations,
            COUNT(e.name) AS number_of_employees
        FROM
            `tabEmployee Grade` eg
        LEFT JOIN
            `tabDesignation` d
            ON d.custom_grade = eg.name
        LEFT JOIN
            `tabEmployee` e
            ON e.designation = d.name
            AND e.status = 'Active'
            AND e.relieving_date IS NULL
        GROUP BY
            eg.name
        ORDER BY
            CAST(SUBSTRING_INDEX(eg.name, '-', -1) AS UNSIGNED) 
    """

    return frappe.db.sql(query, as_dict=True)


# -------------------------------------------------------
# Message
# -------------------------------------------------------

def get_message():
    return """
        <span class="indicator" style="font-weight:bold;font-size:15px;">
            Note :
        </span>
        <span class="indicator blue" style="font-size:15px;">
            Designations are shown grade-wise and separated by commas.
        </span>
    """

File: gke_customization/gke_catalog/report/grade_&_designation_wise_employee_count/grade_&_designation_wise_employee_count.js
// Copyright (c) 2025, Gurukrupa Export and contributors
// For license information, please see license.txt

frappe.query_reports["Grade & Designation wise Employee count"] = {
	"filters": [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			reqd: 0,
            on_change: function () {
				let company = frappe.query_report.get_filter_value("company");
				let branch_filter = frappe.query_report.get_filter("branch");
				if (company && company.trim().toLowerCase() === "gurukrupa export private limited") {
					branch_filter.df.hidden = 0;
					branch_filter.toggle(true);
				} else {
					branch_filter.df.hidden = 1;
					branch_filter.toggle(false);
				}
                frappe.query_report.refresh();
			}
		},
		{
			fieldname: "branch",
			label: __("Branch"),
			fieldtype: "MultiSelectList",
			options: "Branch",
			reqd: 0,
            hidden:1,
			get_data: function (txt) {
				return frappe.db.get_link_options("Branch", txt);
			},
            on_change: () => frappe.query_report.refresh()
		},

	],


onload: function(report) {
    if (frappe.query_report.message) {
            const note = frappe.query_report.message;
            frappe.msgprint({
                message: note,
                indicator: 'blue'
            });
        } 

        let company = report.get_filter_value("company");
		let branch_filter = report.get_filter("branch");
		if (company && company.trim().toLowerCase() === "gurukrupa export private limited") {
			branch_filter.df.hidden = 0;
			branch_filter.toggle(true);
		} else {
			branch_filter.df.hidden = 1;
			branch_filter.toggle(false);
		}
    
        function fetchOptions(doctype, field, filterField, includeBlank = false) {
            frappe.call({
                method: "frappe.client.get_list",
                args: {
                    doctype: doctype,
                    fields: [`distinct ${field}`],
                    order_by: `${field} asc`,
                    limit_page_length: 30000,
                },
                callback: function (r) {
                    if (r.message) {
                        let options = r.message
                            .map(row => row[field])
                            .filter(value => value && value.trim() !== "");
                        if (includeBlank) {
                            options.unshift(""); 
                        }
                        const filter = report.get_filter(filterField);
                        filter.df.options = options;
                        filter.refresh();
                    }
                },
            });
        }
    
      fetchOptions("Employee", "manufacturer", "manufacturer",true);
    //   fetchOptions("Employee", "custom_operation", "operation",true);

	

       
        
    
        report.page.add_inner_button(__("Clear Filter"), function () {
            report.filters.forEach(function (filter) {
                let field = report.get_filter(filter.fieldname);
                if (field.df.fieldtype === "MultiSelectList") {
                    field.set_value([]);
                } else if (field.df.default) {
                    field.set_value(field.df.default);
                } else {
                    field.set_value("");
                }
            });
        });
        
    }}
    



		

